src/lib/utils/SynapseClient.tsx

```diff
@@ -608,3 +638,245 @@ export const signOut = () => {
+    })
+  })
+}
+
+const calculateMd5 = (
+  fileBlob: File | Blob
+) => {
+  const bmf = new browserMd5File()
+  return new Promise((resolve, reject) => {
+    bmf.md5(
+      fileBlob,
+      (error: any, md5: string) => {
+        if (md5) {
+          resolve(md5)
+        } else if (error) {
+          reject(error)
+        }
+      },
+      (progress: number) => {
+        // console.log('progress: ', progress)
+      }
+    )
+  })
+}
+
+const processFilePart = (
+  partNumber: number,
+  multipartUploadStatus: MultipartUploadStatus,
+  sessionToken: string | undefined,
+  fileName: string,
+  file: Blob,
+  request: MultipartUploadRequest,
+  fileUploadResolve: (fileUpload: FileUploadComplete) => void,
+  fileUploadReject: (reason: any) => void,
+  endpoint: string = DEFAULT_ENDPOINT
+) => {
+  if (multipartUploadStatus.clientSidePartsState![partNumber - 1]) {
+    // no-op. this part has already been processed!
+    return
+  }
+
+  const uploadID = multipartUploadStatus.uploadId
+  const presignedUploadUrlRequest: BatchPresignedUploadUrlRequest = {
+    contentType: 'application/octet-stream', // each part is binary
+    partNumbers: [partNumber],
+    uploadId: uploadID
+  }
+  const presignedUrlUrl = `/file/v1/file/multipart/${uploadID}/presigned/url/batch`
+  doPost(presignedUrlUrl, presignedUploadUrlRequest, sessionToken, undefined, endpoint).then(
+    async (presignedUrlResponse: BatchPresignedUploadUrlResponse) => {
+      const presignedUrl = presignedUrlResponse.partPresignedUrls[0].uploadPresignedUrl
+      // calculate the byte range
+      const startByte = (partNumber - 1) * request.partSizeBytes
+      let endByte = partNumber * request.partSizeBytes - 1
+      if (endByte >= request.fileSizeBytes) {
+        endByte = request.fileSizeBytes - 1
+      }
+      const fileSlice = file.slice(startByte, endByte + 1, presignedUploadUrlRequest.contentType)
+      await uploadFilePart(presignedUrl, fileSlice, presignedUploadUrlRequest.contentType)
+      // uploaded the part.  calculate md5 of the part and add the part to the upload
+      calculateMd5(fileSlice).then((md5: string) => {
+        const addPartUrl = `/file/v1/file/multipart/${uploadID}/add/${partNumber}?partMD5Hex=${md5}`
+        doPut(addPartUrl, undefined, sessionToken, undefined, endpoint).then(
+          (addPartResponse: AddPartResponse) => {
+            if (addPartResponse.addPartState === 'ADD_SUCCESS') {
+              // done with this part!
+              multipartUploadStatus.clientSidePartsState![partNumber - 1] = true
+              checkUploadComplete(
+                multipartUploadStatus,
+                fileName,
+                sessionToken,
+                fileUploadResolve,
+                fileUploadReject,
+                endpoint)
+            } else {
+              // retry after a brief delay
+              delay(1000).then(() => {
```

Comment:
Does this ever fail entirely? I'm sure there's lots of edge cases to handle, but this seems like one of the more fragile situations,  a TODO: Handle if filepart doesn't upload after x number of attempts could be good

Reply:
Could have a max limit, but in general we just want to keep trying until it works.